---
title: "R Notebook"
output: html_notebook
---

Rscript -e "rmarkdown::render('correlation_analysis_consensus_traits.Rmd')"

Loading packages
```{r loading packages}
library(wrapr)
library(data.table)
library(pcaMethods)
```


Loading functions for matrix-matrix correlation analysis
```{r loading functions for matrix-matrix correlation analysis}
# get all indices of a matrix
# for symetric matrix size is a number otherwise size=(row, col)
getMatrixIndex = function(size, symmetric=T, diagonal=F)
{
   allidx = NULL

   if(symmetric){
      for(i in c(1:(size[1]-1)) ) {
         iv = cbind(i, (i+1):size[1])
         allidx = rbind(allidx, iv)
      }
      if(diagonal) {allidx = rbind(allidx, cbind(1:size[1], 1:size[1]) )}

   } else {
      for(i in c(1:(size[1])) ) {
         iv = cbind(i, 1:(size[2]))
         allidx = rbind(allidx, iv)
      }
   }

   return (allidx)
}


simpleCorTest=function(x,y){
signif( cor.test(x,y,method="p",use="p")$p.value ,2) 
}

# no of rows of amatrix is the same as the length of myvect
corTest4multivects=function(myvect, amatrix){
pvals = apply(amatrix, 2, simpleCorTest, y=myvect)
#cat(pvals[], "\n")
as.numeric(pvals)
}

# compute correlation coefficients (spearman, pearson), pvalues of the columns 
corRhoPvalSpearmanPearson = function (datMatrix) {

  rho=cor(datMatrix, method="pearson", use="complete.obs")
  pval=apply(datMatrix, 2, corTest4multivects, datMatrix)

  #datMatrixRank = apply(datMatrix, 2, rank)
  rhoR=cor(datMatrix, method="spearman", use="complete.obs")
  #pvalR=apply(datMatrixRank, 2, corTest4multivects, xdatMatrixRank)

  midx = getMatrixIndex(size=dim(rho), symmetric=TRUE, diagonal=FALSE)
  id1  = colnames(datMatrix)[midx[,1]]
  corMatrix = cbind(colnames(datMatrix)[midx[,1]], colnames(datMatrix)[midx[,2]],
              signif(rho[midx],3),signif(rhoR[midx],3), signif(pval[midx],3))
  colnames(corMatrix) = c("TraitA", "TraitB", "rho_pearson", "rho_spearman", "pvalue")

  return (corMatrix)

}


# compute correlation coefficients (spearman, pearson), pvalues of the columns 
corRhoPvalSpearmanPearson_TwoMatrices = function (datMatrix, datMatrix2) {

  rho=cor(datMatrix, datMatrix2, method="pearson", use="complete.obs")
  pval=apply(datMatrix, 2, corTest4multivects, datMatrix2)
  pval=t(pval)

  datMatrixRank = apply(datMatrix, 2, rank)
  datMatrixRank = matrix(as.integer(datMatrixRank), nrow=nrow(datMatrixRank))
  #datMatrixRankT = t(datMatrixRank)

  datMatrixRank2 = apply(datMatrix2, 2, rank)
  datMatrixRank2 = matrix(as.integer(datMatrixRank2), nrow=nrow(datMatrixRank2))
  #datMatrixRankT2 = t(datMatrixRank2)

  rhoR=cor(datMatrixRank, datMatrixRank2, method="pearson", use="complete.obs")
  pvalR=apply(datMatrixRank, 2, corTest4multivects, datMatrixRank2)
  pvalR=t(pvalR)

  midx = getMatrixIndex(size=dim(rho), symmetric=FALSE, diagonal=TRUE)
  id1  = colnames(datMatrix)[midx[,1]]
  corMatrix = cbind(colnames(datMatrix)[midx[,1]], colnames(datMatrix2)[midx[,2]],
              signif(rho[midx],3),signif(pval[midx],3), signif(rhoR[midx],3), signif(pvalR[midx],3))
  colnames(corMatrix) = c("TraitA", "TraitB", "rho_pearson", "pvalue_pearson", "rho_spearman", "pvalue_spearman")

  return (corMatrix)

}
```


Reading gene expression values for brain regions
```{r Reading gene expression values for 8 brain regions}
# BM10
expr_bm10 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.BM_10.PMI_race_sex_RIN_exonicRate_rRnaRate_batch_adj.tsv", data.table = F)
annot_bm <- expr_bm10[, c(1,2)]
row.names(expr_bm10) <- expr_bm10[[1]]
expr_bm10 <- expr_bm10[, -c(1, 2)]
# BM22
expr_bm22 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.BM_22.PMI_race_sex_RIN_exonicRate_rRnaRate_batch_adj.tsv", data.table = F)
row.names(expr_bm22) <- expr_bm22[[1]]
expr_bm22 <- expr_bm22[, -c(1, 2)]
# BM36
expr_bm36 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.BM_36.PMI_race_sex_RIN_exonicRate_rRnaRate_batch_adj.tsv", data.table = F)
row.names(expr_bm36) <- expr_bm36[[1]]
expr_bm36 <- expr_bm36[, -c(1, 2)]
# BM44
expr_bm44 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.BM_44.PMI_race_sex_RIN_exonicRate_rRnaRate_batch_adj.tsv", data.table = F)
row.names(expr_bm44) <- expr_bm44[[1]]
expr_bm44 <- expr_bm44[, -c(1, 2)]
# ROSMAP
expr_rosmap <- fread("/sc/orga/projects/zhangb03a/shared/ROSMAP/Synapse/RNAseq/covariates_adjustment/expression.log2FPKM.batch_pmi_msex_RIN_adj.no_outliers.symbol_clean.tsv", data.table = F)
annot_ros <- expr_rosmap[, c(1,2)]
row.names(expr_rosmap) <- expr_rosmap[[1]]
expr_rosmap <- expr_rosmap[, -c(1, 2)]
```


Reading meta for brain regions
```{r Reading meta for brain regions}
# BM10
meta_bm10 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.meta.BM_10.tsv")
ceradsc <- setNames(1:4, qc(defAD, probAD, possAD, NL))
meta_bm10$ceradsc <- ceradsc[meta_bm10$CERAD]
#BM22
meta_bm22 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.meta.BM_22.tsv")
meta_bm22$ceradsc <- ceradsc[meta_bm22$CERAD]
#BM36
meta_bm36 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.meta.BM_36.tsv")
meta_bm36$ceradsc <- ceradsc[meta_bm36$CERAD]
#BM22
meta_bm44 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.meta.BM_44.tsv")
meta_bm44$ceradsc <- ceradsc[meta_bm44$CERAD]
# ROSMAP
meta_rosmap <- fread("/sc/orga/projects/zhangb03a/shared/ROSMAP/Synapse/RNAseq/covariates_adjustment/expression.log2FPKM.meta.tsv", skip = 1, drop = 1:2)
colnames(meta_rosmap) <- qc(projid,study,age_first_ad_dx,age_death,age_at_visit_max,pmi,msex,educ,apoe_genotype,race,spanish,braaksc,ad_reagan,cts_mmse30_first_ad_dx,cts_mmse30_lv,ceradsc,cogdx,Rnaseqid,Batch,RINcontinuous,Sampleid,CERAD,Dx.by.braak.cerad)
meta_rosmap$projid <- paste0("Proj", meta_rosmap$projid)
colnames(meta_rosmap)[4] <- "AOD"
meta_rosmap$AOD <- as.numeric(gsub("\\+", "", meta_rosmap$AOD))
meta_rosmap$CERAD <- gsub("DefiniteAD", "defAD", meta_rosmap$CERAD)
```



Checking if sample orders are same between expr and meta
```{r Checking if sample orders are same between expr and meta}
meta_ls <- list(meta_bm10, meta_bm22, meta_bm36, meta_bm44, meta_rosmap)
expr_ls <- list(expr_bm10, expr_bm22, expr_bm36, expr_bm44, expr_rosmap)


mapply(function(x, y) identical(x[[1]], colnames(y)), meta_ls, expr_ls)
```


Reading ifitm3 consensus network
```{r Reading ifitm3 consensus network}
ifitm3_consensus <- fread("/sc/orga/projects/zhangb03a/lei_guo/IFITM3/Corr_Results/Genes_sig_cor_IFITM3_consensus.tsv")

ifitm3_consensus_8 <- ifitm3_consensus[Consensus_Network == "Consensus_all_8"]
ifitm3_consensus_8$Geneid <- annot_bm[[1]][match(ifitm3_consensus_8[[1]], annot_bm[[2]])]

# extracting gene expression values for ifitm3_consensus_8
expr_consensus_8_ls <- lapply(expr_ls, function(x) x[ifitm3_consensus_8[[3]], ])
```


Subsetting AD traits
```{r Subsetting AD traits}
# msbb
traits_msbb <- lapply(meta_ls[1:4], function(x) x[, c("AOD", "CDR", "PlaqueMean", "bbscore", "ceradsc"), with = F])

# rosmap
traits_rosmap <- meta_rosmap[, c("AOD", "braaksc", "cogdx", "ceradsc"), with = F]


expr_consensus_8_ls_t <- lapply(expr_consensus_8_ls, t)
```


gene trait correlation
```{r gene trait correlation}
# rosmap
cor_res_ros <- corRhoPvalSpearmanPearson_TwoMatrices(expr_consensus_8_ls_t[[5]], traits_rosmap)

# adjusting p values
cor_res_ros <- cbind(cor_res_ros, p.adjust(as.numeric(cor_res_ros[, "pvalue_pearson"]), method = "BH"))
cor_res_ros <- cbind(cor_res_ros, p.adjust(as.numeric(cor_res_ros[, "pvalue_spearman"]), method = "BH"), annot_ros[[2]][match(cor_res_ros[, 1], annot_ros[[1]])])
colnames(cor_res_ros)[7:9] <- c("pvalue_pearson_adj", "pvalue_spearman_adj", "GeneSymbol")


# msbb
cor_res_bm <- mapply(corRhoPvalSpearmanPearson_TwoMatrices, expr_consensus_8_ls_t[1:4], traits_msbb, SIMPLIFY = F)

# adjusting p value for cor_res
# spearman
p_adj_spearman_bm <- lapply(cor_res_bm, function(x) p.adjust(as.numeric(x[, "pvalue_spearman"]), method = "BH"))
#pearson
p_adj_pearson_bm <- lapply(cor_res_bm, function(x) p.adjust(as.numeric(x[, "pvalue_pearson"]), method = "BH"))

# adding p.adj column to each result
cor_res_bm <- lapply(cor_res_bm, as.data.frame)
# spearman
cor_res_bm <- mapply(function(x, y) "[<-"(x, "p.adj_spearman_bm", value = y) , cor_res_bm, p_adj_spearman_bm, SIMPLIFY = FALSE)
#pearson
cor_res_bm <- mapply(function(x, y) "[<-"(x, "p.adj_pearson_bm", value = y) , cor_res_bm, p_adj_pearson_bm, SIMPLIFY = FALSE)

cor_res_bm <- mapply(function(x, y) "[<-"(x, "GeneSymbol", value = y) , cor_res_bm, lapply(cor_res_bm, function(x) annot_bm[[2]][match(x[, 1], annot_bm[[1]])]), SIMPLIFY = FALSE)

# combining the results
cor_res_bm_ros <- c(cor_res_bm, list(cor_res_ros))
cor_res_bm_ros <- lapply(cor_res_bm_ros, as.data.frame)
```



Consensus Network trait correlation
```{r Consensus Network trait correlation}
# PCA prep
expr_consensus_8_ls_t_prep <- lapply(expr_consensus_8_ls_t, prep, scale='uv', center=TRUE)
#PCA
expr_consensus_8_pcs <- lapply(expr_consensus_8_ls_t_prep, function(x) pca(x, method='ppca', nPcs=3)@scores)

# network trait correlation
#rosmap
net_trait_cor_ros <- corRhoPvalSpearmanPearson_TwoMatrices(expr_consensus_8_pcs[[5]], traits_rosmap)
# adjusting p values
net_trait_cor_ros <- cbind(net_trait_cor_ros, p.adjust(as.numeric(net_trait_cor_ros[, "pvalue_pearson"]), method = "BH"))
net_trait_cor_ros <- cbind(net_trait_cor_ros, p.adjust(as.numeric(net_trait_cor_ros[, "pvalue_spearman"]), method = "BH"))
colnames(net_trait_cor_ros)[7:8] <- c("pvalue_pearson_adj", "pvalue_spearman_adj")

# msbb
net_trait_cor_bm <- mapply(corRhoPvalSpearmanPearson_TwoMatrices, expr_consensus_8_pcs[1:4], traits_msbb, SIMPLIFY = F)
# adjusting p value for cor_res
# spearman
p_adj_spearman_bm <- lapply(net_trait_cor_bm, function(x) p.adjust(as.numeric(x[, "pvalue_spearman"]), method = "BH"))
#pearson
p_adj_pearson_bm <- lapply(net_trait_cor_bm, function(x) p.adjust(as.numeric(x[, "pvalue_pearson"]), method = "BH"))

# adding p.adj column to each result
net_trait_cor_bm <- lapply(net_trait_cor_bm, as.data.frame)
# spearman
net_trait_cor_bm <- mapply(function(x, y) "[<-"(x, "p.adj_spearman_bm", value = y) , net_trait_cor_bm, p_adj_spearman_bm, SIMPLIFY = FALSE)
#pearson
net_trait_cor_bm <- mapply(function(x, y) "[<-"(x, "p.adj_pearson_bm", value = y) , net_trait_cor_bm, p_adj_pearson_bm, SIMPLIFY = FALSE)

# combining the results
cor_res_net_bm_ros <- c(net_trait_cor_bm, list(net_trait_cor_ros))
cor_res_net_bm_ros <- lapply(cor_res_net_bm_ros, as.data.frame)
```


Writing results
```{r Writing results}
# Consensus gene - trait
output_names_gene_trait <- paste0("/sc/orga/projects/zhangb03a/lei_guo/IFITM3/Corr_Results/IFITM3_consensus_8_trait_cor/", c("consensus_8_gene_trait_cor_test_bm10.csv", "consensus_8_gene_trait_cor_test_bm22.csv", "consensus_8_gene_trait_cor_test_bm36.csv", "consensus_8_gene_trait_cor_test_bm44.csv", "consensus_8_gene_trait_cor_test_rosmap.csv"))
invisible(mapply(fwrite, cor_res_bm_ros, output_names_gene_trait))


# Consensus network - trait
output_names_net_trait <- paste0("/sc/orga/projects/zhangb03a/lei_guo/IFITM3/Corr_Results/IFITM3_consensus_8_trait_cor/", c("consensus_8_trait_cor_test_bm10.csv", "consensus_8_trait_cor_test_bm22.csv", "consensus_8_trait_cor_test_bm36.csv", "consensus_8_trait_cor_test_bm44.csv", "consensus_8_trait_cor_test_rosmap.csv"))
invisible(mapply(fwrite, cor_res_net_bm_ros, output_names_net_trait))
```

