---
title: "R Notebook"
output: html_notebook
---


Rscript -e "rmarkdown::render('make_matrix_for_enrichment.Rmd')"
```{r Loading packages}
library(data.table)
```


Reading correlation results
```{r Reading correlation results}
cor_res_filename <- paste0("/sc/orga/projects/zhangb03a/lei_guo/IFITM3/Corr_Results/", list.files("/sc/orga/projects/zhangb03a/lei_guo/IFITM3/Corr_Results/", pattern = "IFITM3_gene_cor_test_"))

cor_res <- lapply(cor_res_filename, fread)
names(cor_res) <- basename(cor_res_filename)

# getting common 

# subsetting significant results
cor_res_sig <- lapply(cor_res, function(x) x[p.adj_spearman < 0.05])

# seperating cor_res by dataset
cor_res_sig_hb <- cor_res_sig[5:7]
cor_res_sig_hb1 <- lapply(cor_res_sig_hb, function(x) x[TraitA == 10025912066])
cor_res_sig_hb2 <- lapply(cor_res_sig_hb, function(x) x[TraitA == 10033668462])                 


cor_res_sig_bm <- cor_res_sig[1:4]
cor_res_sig_ros <- cor_res_sig[[8]]

# seperating cor_res by +/_ correlation
cor_res_sig_bm_pos <- lapply(cor_res_sig_bm, function(x) x[rho_spearman > 0])
cor_res_sig_bm_neg <- lapply(cor_res_sig_bm, function(x) x[rho_spearman < 0])

cor_res_sig_ros_pos <- cor_res_sig_ros[rho_spearman > 0]
cor_res_sig_ros_neg<- cor_res_sig_ros[rho_spearman < 0]

cor_res_sig_hb1_pos <- lapply(cor_res_sig_hb1, function(x) x[rho_spearman > 0])
cor_res_sig_hb1_neg <- lapply(cor_res_sig_hb1, function(x) x[rho_spearman < 0])

cor_res_sig_hb2_pos <- lapply(cor_res_sig_hb2, function(x) x[rho_spearman > 0])
cor_res_sig_hb2_neg <- lapply(cor_res_sig_hb2, function(x) x[rho_spearman < 0])
```

Reading gene expression values for 8 brain regions
```{r Reading gene expression values for 8 brain regions}
# BM10
expr_bm10 <- fread("/sc/orga/projects/zhangb03a/shared/msbb-seq/Data/mixed_model_correction_updated/msbb.BM_10.PMI_AOD_race_sex_RIN_exonicRate_rRnaRate_batch_adj.tsv")
# ROSMAP
expr_rosmap <- fread("/sc/orga/projects/zhangb03a/shared/ROSMAP/Synapse/RNAseq/covariates_adjustment/expression.log2FPKM.batch_pmi_msex_RIN_adj.no_outliers.symbol_clean.tsv")
# HBTRC CR AD sample
expr_hb_rc <- fread("/sc/orga/projects/zhangb03a/shared/HBTRC/hbtrc_expression_data/xhbtrc-Alzh_CR_AgeCorrected_all.xls")
```

Making GeneID-Symbol annotation table for each dataset
```{r Making GeneID-Symbol annotation table for each dataset}
# msbb
annot_msbb <- expr_bm10[, 1:2]

# rosmap
annot_rosmap <- expr_rosmap[, 1:2]

# hb
annot_hb <- expr_hb_rc[, c(1,3)]
colnames(annot_hb) <- c("Geneid", "Symbol")
annot_hb$Geneid <- as.character(annot_hb$Geneid)
```

Annotating Geneid
```{r Annotating Geneid}
# msbb
cor_res_sig_bm_sym <- lapply(cor_res_sig_bm, function(x) unique(annot_msbb$Symbol[match(x$TraitB, annot_msbb$Geneid)]))
mod_bm <- mapply(rep, c("IFITM3_BM10", "IFITM3_BM22", "IFITM3_BM36", "IFITM3_BM44"), lengths(cor_res_sig_bm_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_bm <- mapply(cbind, cor_res_sig_bm_sym, mod_bm)
IFITM3_mx_bm <- lapply(IFITM3_mx_bm, as.data.frame)

cor_res_sig_bm_pos_sym <- lapply(cor_res_sig_bm_pos, function(x) unique(annot_msbb$Symbol[match(x$TraitB, annot_msbb$Geneid)]))
mod_bm_pos <- mapply(rep, c("IFITM3_bm10_pos", "IFITM3_bm22_pos", "IFITM3_bm36_pos", "IFITM3_bm44_pos"), lengths(cor_res_sig_bm_pos_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_bm_pos <- mapply(cbind, cor_res_sig_bm_pos_sym, mod_bm_pos)
IFITM3_mx_bm_pos <- lapply(IFITM3_mx_bm_pos, as.data.frame)

cor_res_sig_bm_neg_sym <- lapply(cor_res_sig_bm_neg, function(x) unique(annot_msbb$Symbol[match(x$TraitB, annot_msbb$Geneid)]))
mod_bm_neg <- mapply(rep, c("IFITM3_bm10_neg", "IFITM3_bm22_neg", "IFITM3_bm36_neg", "IFITM3_bm44_neg"), lengths(cor_res_sig_bm_neg_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_bm_neg <- mapply(cbind, cor_res_sig_bm_neg_sym, mod_bm_neg)
IFITM3_mx_bm_neg <- lapply(IFITM3_mx_bm_neg, as.data.frame)

# rosmap
cor_res_sig_ros_sym <- unique(annot_rosmap$Symbol[match(cor_res_sig_ros$TraitB, annot_rosmap$Geneid)])
mod_ros <- rep("IFITM3_ROSMAP", length(cor_res_sig_ros_sym))
IFITM3_mx_ros <- as.data.frame(cbind(cor_res_sig_ros_sym, mod_ros))
colnames(IFITM3_mx_ros) <- c("V1", "V2")

cor_res_sig_ros_pos_sym <- unique(annot_rosmap$Symbol[match(cor_res_sig_ros_pos$TraitB, annot_rosmap$Geneid)])
mod_ros_pos <- rep("IFITM3_ROSMAP_pos", length(cor_res_sig_ros_pos_sym))
IFITM3_mx_ros_pos <- as.data.frame(cbind(cor_res_sig_ros_pos_sym, mod_ros_pos))
colnames(IFITM3_mx_ros_pos) <- c("V1", "V2")

cor_res_sig_ros_neg_sym <- unique(annot_rosmap$Symbol[match(cor_res_sig_ros_neg$TraitB, annot_rosmap$Geneid)])
mod_ros_neg <- rep("IFITM3_ROSMAP_neg", length(cor_res_sig_ros_neg_sym))
IFITM3_mx_ros_neg <- as.data.frame(cbind(cor_res_sig_ros_neg_sym, mod_ros_neg))
colnames(IFITM3_mx_ros_neg) <- c("V1", "V2")

# hb1
cor_res_sig_hb1_sym <- lapply(cor_res_sig_hb1, function(x) unique(annot_hb$Symbol[match(as.character(x$TraitB), annot_hb$Geneid)]))
mod_hb1 <- mapply(rep, paste0(annot_hb$Symbol[match("10023817038", annot_hb$Geneid)], "_V1_", c("HBCR", "HBPFC", "HBVC")), lengths(cor_res_sig_hb1_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_hb1 <- mapply(cbind, cor_res_sig_hb1_sym, mod_hb1)
IFITM3_mx_hb1 <- lapply(IFITM3_mx_hb1, as.data.frame)

cor_res_sig_hb1_pos_sym <- lapply(cor_res_sig_hb1_pos, function(x) unique(annot_hb$Symbol[match(as.character(x$TraitB), annot_hb$Geneid)]))
mod_hb1_pos <- mapply(rep, paste0(annot_hb$Symbol[match("10023817038", annot_hb$Geneid)], "_V1_", c("HBCR_pos", "HBPFC_pos", "HBVC_pos")), lengths(cor_res_sig_hb1_pos_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_hb1_pos <- mapply(cbind, cor_res_sig_hb1_pos_sym, mod_hb1_pos)
IFITM3_mx_hb1_pos <- lapply(IFITM3_mx_hb1_pos, as.data.frame)

cor_res_sig_hb1_neg_sym <- lapply(cor_res_sig_hb1_neg, function(x) unique(annot_hb$Symbol[match(as.character(x$TraitB), annot_hb$Geneid)]))
mod_hb1_neg <- mapply(rep, paste0(annot_hb$Symbol[match("10023817038", annot_hb$Geneid)], "_V1_", c("HBCR_neg", "HBPFC_neg", "HBVC_neg")), lengths(cor_res_sig_hb1_neg_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_hb1_neg <- mapply(cbind, cor_res_sig_hb1_neg_sym, mod_hb1_neg)
IFITM3_mx_hb1_neg <- lapply(IFITM3_mx_hb1_neg, as.data.frame)

# hb2
cor_res_sig_hb2_sym <- lapply(cor_res_sig_hb2, function(x) unique(annot_hb$Symbol[match(as.character(x$TraitB), annot_hb$Geneid)]))
mod_hb2 <- mapply(rep, paste0(annot_hb$Symbol[match("10025903125", annot_hb$Geneid)], "_V2_", c("HBCR", "HBPFC", "HBVC")), lengths(cor_res_sig_hb2_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_hb2 <- mapply(cbind, cor_res_sig_hb2_sym, mod_hb2)
IFITM3_mx_hb2 <- lapply(IFITM3_mx_hb2, as.data.frame)

cor_res_sig_hb2_pos_sym <- lapply(cor_res_sig_hb2_pos, function(x) unique(annot_hb$Symbol[match(as.character(x$TraitB), annot_hb$Geneid)]))
mod_hb2_pos <- mapply(rep, paste0(annot_hb$Symbol[match("10023817038", annot_hb$Geneid)], "_V1_", c("HBCR_pos", "HBPFC_pos", "HBVC_pos")), lengths(cor_res_sig_hb2_pos_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_hb2_pos <- mapply(cbind, cor_res_sig_hb2_pos_sym, mod_hb2_pos)
IFITM3_mx_hb2_pos <- lapply(IFITM3_mx_hb2_pos, as.data.frame)

cor_res_sig_hb2_neg_sym <- lapply(cor_res_sig_hb2_neg, function(x) unique(annot_hb$Symbol[match(as.character(x$TraitB), annot_hb$Geneid)]))
mod_hb2_neg <- mapply(rep, paste0(annot_hb$Symbol[match("10023817038", annot_hb$Geneid)], "_V1_", c("HBCR_neg", "HBPFC_neg", "HBVC_neg")), lengths(cor_res_sig_hb2_neg_sym), MoreArgs = list(SIMPLIFY = F))
IFITM3_mx_hb2_neg <- mapply(cbind, cor_res_sig_hb2_neg_sym, mod_hb2_neg)
IFITM3_mx_hb2_neg <- lapply(IFITM3_mx_hb2_neg, as.data.frame)

# combine the dfs
IFITM3_df_all <- rbindlist(c(IFITM3_mx_bm, IFITM3_mx_bm_pos, IFITM3_mx_bm_neg,list(IFITM3_mx_ros, IFITM3_mx_ros_pos, IFITM3_mx_ros_neg), IFITM3_mx_hb1, IFITM3_mx_hb1_pos, IFITM3_mx_hb1_neg, IFITM3_mx_hb2, IFITM3_mx_hb2_pos, IFITM3_mx_hb2_neg))
colnames(IFITM3_df_all) <- c("Genes_cor_IFITM3", "Datasets")
```



Writing results
```{r Writing results}
fwrite(IFITM3_df_all, "/sc/orga/projects/zhangb03a/lei_guo/IFITM3/Corr_Results/Genes_sig_cor_IFITM3_8_datasets.tsv", sep = "\t")
```

